src/cache.rs

```diff
@@ -0,0 +1,292 @@
+        self.cache.get( attribute ).map(|m| &m.cache )
+    }
+
+    pub fn get_for_entid(&mut self, attribute: NamespacedKeyword, entid: KnownEntid) -> Option<&Vec<TypedValue>> {
+        if let Some(c) = self.cache.get(&attribute) {
+            c.get(&entid.into())
+        } else { None }
+    }
+}
+
+struct AttributeValueProvider {
+    attribute: NamespacedKeyword,
+}
+
+impl ValueProvider<TypedValue, Vec<TypedValue>> for AttributeValueProvider {
+    fn fetch_values<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema) -> Result<BTreeMap<TypedValue, Vec<TypedValue>>> {
+        Ok(q_once(sqlite,
+                  schema,
+                  r#"[:find ?entity ?value
+                                          :in ?attribute
+                                          :where [?entity ?a        ?value]
+                                                 [?a      :db/ident ?attribute]]"#,
+                  QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?attribute"), self.attribute.clone().into())]))
+            .into_rel_result()?
+            .into_iter()
+            .fold(BTreeMap::new(), |mut map, row| {
+                map.entry(row[0].clone()).or_insert(vec![]).push(row[1].clone());
+                map
+            }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mentat_db::db;
+    use conn::Conn;
+
+    fn assert_values_present_for_attribute(attribute_cache: &mut AttributeCacher, attribute: &NamespacedKeyword, values: Vec<Vec<TypedValue>>) {
+        let cached_values: Vec<Vec<TypedValue>> = attribute_cache.get(&attribute)
+            .expect("Expected cached values")
+            .values()
+            .cloned()
+            .collect();
+        assert_eq!(cached_values, values);
+    }
+
+    #[test]
+    fn test_add_to_cache() {
+        let mut sqlite = db::new_connection("").unwrap();
+        let mut conn = Conn::connect(&mut sqlite).unwrap();
+        let _report = conn.transact(&mut sqlite, r#"[
+            {  :db/ident       :foo/bar
+               :db/valueType   :db.type/long },
+            {  :db/ident       :foo/baz
+               :db/valueType   :db.type/boolean }]"#).expect("transaction expected to succeed");
+        let _report = conn.transact(&mut sqlite, r#"[
+            {  :foo/bar        100
+               :foo/baz        false },
+            {  :foo/bar        200
+               :foo/baz        true }]"#).expect("transaction expected to succeed");
+        let schema = conn.current_schema();
+        let kw = NamespacedKeyword::new("foo", "bar");
+        let mut attribute_cache = AttributeCacher::new();
+        attribute_cache.register_attribute(&sqlite, &schema, kw.clone() ).expect("No errors on add to cache");
+        assert_values_present_for_attribute(&mut attribute_cache, &kw, vec![vec![TypedValue::Long(100)], vec![TypedValue::Long(200)]]);
```

Comment:
Can we get a test for the primary use case of the cache?

* Add a bunch of data
* Ask for an attribute to be cached
* Retrieve the corresponding value for a particular (e, a) pair without touching disk.